File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
   TreeNode* helper(TreeNode* node, TreeNode* p, TreeNode* q){
        if(node == NULL) return NULL;

        if(node == p || node == q) return node;
        TreeNode* left = helper(node->left,p,q);
        TreeNode* right = helper(node->right,p,q);

        if(left && right){
            return node;
        }
        else if(left == NULL){
            return right;
        }
        else{
            return left;
        }
   }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return helper(root,p,q);
    }
};

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    TreeNode ans = null;
    int find(TreeNode root, TreeNode p, TreeNode q){
      
     if(root == null) return 0;
     int left = find(root.left, p , q);
     int right = find(root.right, p , q);
     int self = 0;
     if(root.val == p.val || root.val == q.val){
        self = 1;
     }
     int total = self + left + right;
     if(total == 2 && ans == null) ans = root;
     return total;

    }
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null) return null;
        if(root.left == null && root.right == null) return root;

        find(root, p , q);
        return ans;
    }
}